pkg/cli/cli_options_test.go

```diff
@@ -0,0 +1,185 @@
+package cli
+
+import (
+	"errors"
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v3"
+	"os"
+	"testing"
+)
+
+func TestSetOptions(t *testing.T) {
+	cases := []struct {
+		name         string
+		set          string
+		to           string
+		originalYaml string
+		expectYaml   string
+		expectErr    bool
+		expectWarns  []string
+	}{
+		{
+			name: "valid option",
+			// blank vs non-blank isn't too interesting, because yaml_util handles that. See its tests for details.
+			originalYaml: "",
+			set:          "update.stream",
+			to:           "my-stream",
+			expectYaml:   "update:\n    stream: my-stream\n",
+		},
+		{
+			name:      "invalid option",
+			set:       "update",
+			to:        "this can't be a scalar",
+			expectErr: true,
+		},
+		{
+			name:        "unknown option",
+			set:         "foo.bar",
+			to:          "world",
+			expectYaml:  "foo:\n    bar: world\n",
+			expectWarns: []string{`Unrecognized option "foo.bar". We'll still set it, but it may not have any effect.`},
+		},
+		{
+			name:         "yaml starts with unknown option",
+			set:          "update.stream",
+			to:           "my-stream",
+			originalYaml: "extra: option",
+			expectYaml:   "extra: option\nupdate:\n    stream: my-stream\n",
+			expectWarns: []string{
+				`Existing options contain extra parameters:`,
+				`▸ line 1: field extra not found in type cli.Options`,
+			},
+		},
+		{
+			name:         "yaml starts invalid",
+			set:          "update.stream",
+			to:           "my-stream",
+			originalYaml: "update: this can't be a scalar",
+			expectErr:    true,
+		},
+	}
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+			assert := assert.New(t)
+			logger := mockedLogger{}
+			options := map[string]string{
+				tt.set: tt.to,
+			}
+
+			result, err := setOptions([]byte(tt.originalYaml), options, &logger)
+			assert.Equal(tt.expectWarns, logger.warns)
+			assert.Equal(tt.expectErr, err != nil)
+			if err != nil {
+				return
+			}
+			assert.Equal(tt.expectYaml, string(result))
+		})
+	}
+}
+
+type mockedLogger struct {
+	warns []string
+}
+
+func (ml *mockedLogger) Warn(msg string, fields ...zap.Field) {
```

Comment:
https://github.com/klothoplatform/klotho/blob/main/pkg/validation/validation_test.go#L503

I use a zap observer to test log output if you think its less ugly. I think its personally a bit cleaner than this implementation, but up to you